### ISSUE TITLE
(0, fn)(...args) type of calls

### ISSUE THREAD

**Issue by mr-cheater** [j4k0xb/webcrack#6]
Is there any particular reason to not replace calls like `(0, fn)(...args)` to just `fn(...args)`? Also, would be great to know in which cases webpack does this kind of transformation if anyone knows...

**Comment by j4k0xb** [comment]
It's possible but not guaranteed to be safe
webpack does this so exported functions aren't called with a different `this`
E.g. `this` inside `(0, lib.fn)(arg)` is undefined but in `lib.fn(arg)` it's `lib`

Maybe I can analyze it or just add an option to use unsafe transforms

**Comment by mr-cheater** [comment]
Yeah that's what it does but do you really see any possible issue with not removing context for such function calls? I mean it could be useful in super specific and advanced JS cases, but assuming Webpack usually works with ES/TS codebases, it shouldn't really be the issue. That's why I've asked about cases when it does that as I'm almost sure that it doesn't do this particular transform because of specific of the source code and in raw ES/TS source the calls would be made with `this` context anyway.

**Comment by mr-cheater** [comment]
Seems like webpack does this only in case `fn` has no internal references to `this` so should be perfectly safe to convert it?

**Comment by j4k0xb** [comment]
I think it does it for all functions
Here's an example where it would fail without: https://github.com/es-shims/Promise.allSettled/issues/5
That's what I meant with analyzing... Have to make sure the functions *don't* reference `this`

**Comment by 0xdevalias** [comment]
I came across another tool today that seemed to handle this pretty well:

- https://github.com/pionxzh/wakaru
  - https://wakaru.vercel.app/

Digging through the code a little lead me to this:

- https://github.com/pionxzh/wakaru/blob/main/packages/unminify/README.md#un-indirect-call
- https://github.com/pionxzh/wakaru/blob/main/packages/unminify/src/transformations/un-indirect-call.ts
- https://github.com/pionxzh/wakaru/blob/main/packages/unminify/src/transformations/__tests__/un-indirect-call.spec.ts

**Comment by j4k0xb** [comment]
good idea to do it like this 👍🏻 
Another thing I noticed is that bundlers only create `(0, lib.fn)()` for esm exports, so it should be enough to identify them and convert all `require` to named imports
